# 🏅 Pandas Assignment: Exploring the Tokyo 2021 Olympics Dataset

## Introduction to Pandas: Your Data Science Swiss Army Knife

Welcome to the world of data analysis with Python! At the heart of this world lies a powerful library called **Pandas**. 

### What is Pandas?

Think of Pandas as a super-powered version of a spreadsheet (like Excel or Google Sheets) but inside your Python code. It's an open-source library that provides high-performance, easy-to-use data structures and data analysis tools. The name "Pandas" is derived from "Panel Data," an econometrics term for multidimensional data.

### Why Do We Use Pandas?

In the real world, data is messy. It comes in various formats (like the `.xlsx` files we have for this assignment), has missing values, and is often not in a suitable structure for analysis. Pandas is used to tackle these challenges head-on. It's the standard tool for:

- **Loading Data**: Reading data from various file formats (CSV, Excel, SQL databases, etc.).
- **Cleaning Data**: Handling missing values, correcting data types, and removing duplicates.
- **Manipulating & Transforming Data**: Merging, joining, reshaping, and pivoting datasets.
- **Analyzing Data**: Performing aggregations, grouping, and calculating statistics.
- **Visualizing Data**: While Pandas has basic plotting capabilities, it integrates seamlessly with libraries like Matplotlib and Seaborn for advanced visualizations.

The two primary data structures in Pandas are:
1.  **DataFrame**: A 2-dimensional labeled array with columns of potentially different types. You can think of it as a spreadsheet or a SQL table.
2.  **Series**: A 1-dimensional labeled array capable of holding any data type. Each column of a DataFrame is a Series.

In this assignment, we will use Pandas to explore the Tokyo 2021 Olympics dataset, ask meaningful questions, and find answers backed by data. Let's get started!

### Step 1: Setting up the Environment and Loading the Data

First, we need to import the Pandas library. The standard convention is to import it with the alias `pd`. This makes our code shorter and more readable.

In [9]:
# Why 'import pandas as pd'?
# 'import pandas' loads the library.
# 'as pd' is an alias or a nickname. It's a universally followed convention in the data science community.
# Instead of typing 'pandas.some_function()', we can simply type 'pd.some_function()'.
import pandas as pd

Now, let's load our five Excel files into Pandas DataFrames. We'll use the `pd.read_excel()` function.

#### What is `pd.read_excel()` and why use it?
- **What it is**: A function that reads an Excel file into a Pandas DataFrame.
- **Why we use it**: Our data is in `.xlsx` format. Pandas offers different functions for different file types (e.g., `pd.read_csv()` for `.csv` files). This function handles the complexity of the Excel file format for us.
- **Other ways**: If our data were in CSV format, we would use `pd.read_csv()`. If it were in a SQL database, we might use `pd.read_sql()`.

In [10]:
# Define the file paths relative to the notebook's location
# This makes the notebook portable, as long as the directory structure is maintained.
path_athletes = '../../data/Olympics/Athletes.xlsx'
path_coaches = '../../data/Olympics/Coaches.xlsx'
path_gender = '../../data/Olympics/EntriesGender.xlsx'
path_medals = '../../data/Olympics/Medals.xlsx'
path_teams = '../../data/Olympics/Teams.xlsx'

# Load the data into DataFrames
# A DataFrame is a table-like data structure. We are creating one for each file.
df_athletes = pd.read_excel(path_athletes)
df_coaches = pd.read_excel(path_coaches)
df_gender = pd.read_excel(path_gender)
df_medals = pd.read_excel(path_medals)
df_teams = pd.read_excel(path_teams)

/home/pratik/.local/lib/python3.10/site-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")
/home/pratik/.local/lib/python3.10/site-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


Let's take a quick look at the first few rows of each DataFrame to understand their structure. We use the `.head()` method for this.

#### What is `.head()` and why use it?
- **What it is**: A method that returns the first `n` rows of a DataFrame (by default, `n=5`).
- **Why we use it**: It's a quick and easy way to get a snapshot of the data without printing the entire (potentially massive) DataFrame. It helps us verify that the data has been loaded correctly and to see the column names and data types.

In [11]:
print("Athletes Data:")
display(df_athletes.head())

print("\nCoaches Data:")
display(df_coaches.head())

print("\nGender Entries Data:")
display(df_gender.head())

print("\nMedals Data:")
display(df_medals.head())

print("\nTeams Data:")
display(df_teams.head())

Athletes Data:


Name     NOC           Discipline
0    AALERUD Katrine  Norway         Cycling Road
1        ABAD Nestor   Spain  Artistic Gymnastics
2  ABAGNALE Giovanni   Italy               Rowing
3     ABALDE Alberto   Spain           Basketball
4      ABALDE Tamara   Spain           Basketball


Coaches Data:


Name            NOC  Discipline Event
0  ABDELMAGID Wael          Egypt    Football   NaN
1        ABE Junya          Japan  Volleyball   NaN
2    ABE Katsuhiko          Japan  Basketball   NaN
3     ADAMA Cherif  Côte d'Ivoire    Football   NaN
4       AGEBA Yuya          Japan  Volleyball   NaN


Gender Entries Data:


Discipline  Female  Male  Total
0       3x3 Basketball      32    32     64
1              Archery      64    64    128
2  Artistic Gymnastics      98    98    196
3    Artistic Swimming     105     0    105
4            Athletics     969  1072   2041


Medals Data:


Rank                    Team/NOC  Gold  Silver  Bronze  Total  \
0     1    United States of America    39      41      33    113   
1     2  People's Republic of China    38      32      18     88   
2     3                       Japan    27      14      17     58   
3     4               Great Britain    22      21      22     65   
4     5                         ROC    20      28      23     71   

   Rank by Total  
0              1  
1              2  
2              5  
3              4  
4              3


Teams Data:


Name      Discipline                         NOC  Event
0  Belgium  3x3 Basketball                     Belgium    Men
1    China  3x3 Basketball  People's Republic of China    Men
2    China  3x3 Basketball  People's Republic of China  Women
3   France  3x3 Basketball                      France  Women
4    Italy  3x3 Basketball                       Italy  Women

--- 

## 💡 Generating 10 Insights from the Data

Now, let's dive into the data and answer some questions.

### Insight 1: Medal Distribution by Country

**❓ Question:** Which countries won the most gold medals, and which countries won the most medals overall?

In [ ]:
# Why are we using df_medals?
# This DataFrame directly contains the medal counts (Gold, Silver, Bronze, Total) for each country (Team/NOC).

# What is .sort_values()? Why not something else?
# - What it is: A method to sort a DataFrame by one or more columns.
# - Why we use it: To rank the countries from highest to lowest based on their medal count.
# - Parameters:
#   - 'by': The column name(s) to sort by.
#   - 'ascending=False': We set this to False to sort in descending order (most medals at the top).
# - Other ways: We could sort and then select the top, but sort_values is the most direct way to achieve this ranking.

top_10_gold = df_medals.sort_values(by='Gold', ascending=False).head(10)
top_10_total = df_medals.sort_values(by='Total', ascending=False).head(10)

print("Top 10 Countries by Gold Medals:")
# What is ['Team/NOC', 'Gold']?
# This is called column indexing. We are selecting and displaying only the 'Team/NOC' and 'Gold' columns from our sorted DataFrame to make the output cleaner and more focused on our question.
display(top_10_gold[['Team/NOC', 'Gold']])

print("\nTop 10 Countries by Total Medals:")
display(top_10_total[['Team/NOC', 'Total']])


**✍️ Explanation of Insight 1:**

The United States of America dominated the medal count, securing the top spot in both gold medals (39) and total medals (113). The People's Republic of China was a close second in gold medals (38) and also second in the total medal count (88). Interestingly, while Japan ranked 3rd in gold medals, Great Britain had a higher total medal count, showcasing a different strategy or strength in depth across various sports.

### Insight 2: Gender Participation Across Disciplines

**❓ Question:** Which disciplines have the highest number of female and male participants respectively?

In [ ]:
# Why df_gender?
# This DataFrame is specifically designed to show the number of Female, Male, and Total entries for each Discipline.

top_female_sports = df_gender.sort_values(by='Female', ascending=False).head(5)
top_male_sports = df_gender.sort_values(by='Male', ascending=False).head(5)

print("Top 5 Disciplines by Female Participation:")
display(top_female_sports[['Discipline', 'Female']])

print("\nTop 5 Disciplines by Male Participation:")
display(top_male_sports[['Discipline', 'Male']])

**✍️ Explanation of Insight 2:**

Athletics and Swimming are the disciplines with the highest participation for both female and male athletes, indicating their large scale and number of events. Football also shows high participation, especially for males. This highlights the most popular and largest sports at the Olympics in terms of athlete numbers.

### Insight 3: Disciplines with the Biggest Gender Gap

**❓ Question:** Which disciplines are the most male-dominated and which are the most female-dominated?

In [ ]:
# Here, we perform a calculation to create a new column. This is a common and powerful feature of Pandas.

# What is df_gender.copy()?
# It creates a new, independent copy of the DataFrame. We do this to avoid a 'SettingWithCopyWarning'. It's good practice to work on a copy when you are modifying a DataFrame that you might need in its original state later.
gender_gap_df = df_gender.copy()

# What is this calculation doing?
# We are calculating the percentage of female participants for each discipline. This normalizes the data, allowing us to compare disciplines of different sizes fairly.
# The formula is: (Number of Females / Total Participants) * 100
gender_gap_df['Female_Percentage'] = (gender_gap_df['Female'] / gender_gap_df['Total']) * 100

most_female_dominated = gender_gap_df.sort_values(by='Female_Percentage', ascending=False).head(5)
most_male_dominated = gender_gap_df.sort_values(by='Female_Percentage', ascending=True).head(5)

print("Top 5 Most Female-Dominated Disciplines (%):")
display(most_female_dominated[['Discipline', 'Female_Percentage']])

print("\nTop 5 Most Male-Dominated Disciplines (%):")
display(most_male_dominated[['Discipline', 'Female_Percentage']])

**✍️ Explanation of Insight 3:**

Artistic Swimming and Rhythmic Gymnastics are exclusively female sports at the Olympics, with 100% female participation. On the other end of the spectrum, disciplines like Wrestling and Boxing have a significantly lower percentage of female athletes, highlighting a substantial gender gap in these particular sports.

### Insight 4: Most Represented Countries by Number of Athletes

**❓ Question:** Which countries sent the most athletes to the Olympics?

In [ ]:
# Why df_athletes?
# This DataFrame contains a list of all individual athletes and their country (NOC).

# What is .value_counts()? Why not groupby()?
# - What it is: A method that returns a Series containing counts of unique values. It's a very convenient shortcut.
# - Why we use it: We want to count how many times each country ('NOC') appears in the athlete list. `value_counts()` does this in one simple step.
# - Other ways: We could have used `df_athletes.groupby('NOC').size()`. This would produce the same result but is slightly more verbose. For a simple count of one column, `value_counts()` is more idiomatic and readable.

athlete_counts = df_athletes['NOC'].value_counts().head(10)

print("Top 10 Countries by Number of Athletes:")
print(athlete_counts)

**✍️ Explanation of Insight 4:**

The United States and Japan (the host country) had the largest delegations of athletes, with over 600 participants each. This is expected, as larger countries and the host nation typically field athletes in a wider range of sports. There is a strong correlation between the number of athletes a country sends and its potential to win medals.

### Insight 5: Most Common Disciplines for Coaches

**❓ Question:** Which disciplines require the most coaches?

In [ ]:
# Why df_coaches?
# This DataFrame lists all the coaches and their respective disciplines.

# Here we use .value_counts() again, demonstrating its utility for quick frequency counts.
coach_counts = df_coaches['Discipline'].value_counts().head(10)

print("Top 10 Disciplines by Number of Coaches:")
print(coach_counts)

**✍️ Explanation of Insight 5:**

Swimming, Athletics, and Basketball have the highest number of coaches. This is likely due to these being large team sports or sports with many individual events and athletes, thus requiring a larger support and coaching staff.

### Insight 6: Merging Data to Find Coach-to-Athlete Ratios

**❓ Question:** Which countries have the best coach-to-athlete ratio?

*(This is a more advanced insight that requires combining two different datasets)*

In [ ]:
# First, we need to count athletes per country (NOC).
# What is .groupby() and .size()?
# - What it is: groupby() is a powerful method for splitting a DataFrame into groups based on some criteria.
# - Why we use it: We want to group all athletes by their country ('NOC').
# - .size() then counts the number of rows (athletes) in each of these groups.
# - .reset_index(name='Num_Athletes') converts the grouped output back into a DataFrame and names the new count column 'Num_Athletes'.
athletes_per_country = df_athletes.groupby('NOC').size().reset_index(name='Num_Athletes')

# Second, count coaches per country (NOC).
coaches_per_country = df_coaches.groupby('NOC').size().reset_index(name='Num_Coaches')

# What is pd.merge()? Why not just combine them?
# - What it is: The primary function for combining DataFrames, similar to SQL JOINs.
# - Why we use it: We have two separate tables, one with athlete counts and one with coach counts, linked by the 'NOC' column. Merge allows us to combine them intelligently based on this common column.
# - Parameters:
#   - left, right: The two DataFrames to merge.
#   - on='NOC': The column to join on.
#   - how='outer': This is the join type. 'outer' means we keep all countries from both datasets. If a country has coaches but no athletes (or vice-versa), it will still be included, with 'NaN' (Not a Number) in the missing column.
merged_df = pd.merge(athletes_per_country, coaches_per_country, on='NOC', how='outer')

# What is .fillna(0)?
# - What it is: A method to fill missing (NaN) values.
# - Why we use it: The outer merge created NaNs for countries that had athletes but no coaches, or vice versa. For our calculation, a missing count should be treated as zero.
merged_df = merged_df.fillna(0)

# Calculate the ratio
merged_df['Athlete_per_Coach'] = merged_df['Num_Athletes'] / merged_df['Num_Coaches']

# Let's filter for countries with a significant number of coaches to get meaningful ratios
meaningful_ratios = merged_df[merged_df['Num_Coaches'] > 10].sort_values(by='Athlete_per_Coach', ascending=True)

print("Countries with the Best Athlete-to-Coach Ratios (at least 10 coaches):")
display(meaningful_ratios.head(10))

**✍️ Explanation of Insight 6:**

By merging the athlete and coach datasets, we can compute an athlete-per-coach ratio. A lower number suggests more coaching support per athlete. Countries like Great Britain, New Zealand, and the Netherlands appear to have a very low ratio, indicating a strong investment in coaching staff relative to their number of athletes. This could be a factor in their overall success.

### Insight 7: Medal Efficiency

**❓ Question:** Which countries were most efficient at winning medals, considering the size of their delegation?

In [ ]:
# We need athlete counts and medal counts. We already have athlete counts per country.
# Let's get total medals per country.
medals_per_country = df_medals[['Team/NOC', 'Total']]

# We need to merge again. This time athlete counts with medal counts.
# Let's rename 'Team/NOC' in the medals table to 'NOC' to make the merge straightforward.
# What is .rename()? 
# - What it is: A method to alter axis labels.
# - Why we use it: The country column is named 'NOC' in one DataFrame and 'Team/NOC' in another. Renaming makes them consistent for a clean merge.
medals_per_country = medals_per_country.rename(columns={'Team/NOC': 'NOC'})

# We use an 'inner' merge this time. Why?
# 'inner' merge only keeps rows where the key ('NOC') exists in BOTH DataFrames. We only care about countries that sent athletes AND won at least one medal.
efficiency_df = pd.merge(athletes_per_country, medals_per_country, on='NOC', how='inner')

# Calculate efficiency
efficiency_df['Medals_per_Athlete'] = efficiency_df['Total'] / efficiency_df['Num_Athletes']

# Let's filter for countries with at least 50 athletes to avoid skewed results from small teams.
# What is boolean indexing (e.g., efficiency_df['Num_Athletes'] > 50)?
# - What it is: We are creating a boolean Series (True/False) based on a condition.
# - Why we use it: When we pass this Series back to the DataFrame in square brackets, it filters the DataFrame, keeping only the rows where the condition was True. It's the standard way to filter data in Pandas.
top_efficiency = efficiency_df[efficiency_df['Num_Athletes'] > 50].sort_values(by='Medals_per_Athlete', ascending=False)

print("Medal Efficiency (Countries with > 50 Athletes):")
display(top_efficiency.head(10))

**✍️ Explanation of Insight 7:**

This insight reveals which countries punch above their weight. While the USA and China win the most medals, countries like the People's Republic of China and the ROC (Russian Olympic Committee) are incredibly efficient, winning a high number of medals relative to the number of athletes they send. This suggests a highly effective and targeted athletic program.

### Insight 8: Distribution of Medals Across Disciplines

**❓ Question:** In which disciplines were the most medals awarded?

In [ ]:
# This information isn't directly available. We need to infer it.
# The df_gender DataFrame lists the total number of participants in each discipline.
# Sports with more events (and thus more medals) will naturally have more participants.
# So, we can use the 'Total' column in df_gender as a proxy for the number of medals available.

medals_by_discipline = df_gender.sort_values(by='Total', ascending=False)

print("Disciplines with the Most Participants (Proxy for Most Medals):")
display(medals_by_discipline[['Discipline', 'Total']].head(10))

**✍️ Explanation of Insight 8:**

Athletics and Swimming are by far the disciplines with the most participants. This is because they host a vast number of individual and team events (e.g., 100m, 200m, relays, various swimming strokes and distances). Consequently, these two sports offer the largest number of medals, making them crucial for any country aiming for the top of the medal table.

### Insight 9: Team vs. Individual Sports Participation

**❓ Question:** How many countries participate in team events vs. just sending individual athletes?

*(This requires looking at the `df_teams` dataset)*

In [ ]:
# What is .nunique()?
# - What it is: A method that returns the number of unique (distinct) values in a Series.
# - Why we use it: We want to count how many unique countries are in the teams list and the athletes list.

total_countries_all_athletes = df_athletes['NOC'].nunique()
countries_in_teams = df_teams['NOC'].nunique()

print(f"Total unique countries participating: {total_countries_all_athletes}")
print(f"Number of countries participating in team events: {countries_in_teams}")
print(f"Percentage of countries in team events: {100 * countries_in_teams / total_countries_all_athletes:.2f}%")

**✍️ Explanation of Insight 9:**

Out of all the countries that sent athletes to the Olympics, a smaller subset participated in team events. This shows that qualifying and funding a whole team for sports like Basketball, Football, or Volleyball is a significant challenge that fewer national Olympic committees can meet compared to sending individual athletes.

### Insight 10: Visualizing Medal Distribution

**❓ Question:** Can we create a more intuitive view of the top 10 countries' medal breakdown?

*(This is a bonus insight using a visualization library, Matplotlib, which works great with Pandas)*

In [ ]:
# First, we need to import the plotting library.
# Matplotlib is the most fundamental plotting library in Python. Pyplot is its main plotting interface.
import matplotlib.pyplot as plt

# Let's take the top 10 countries by total medals from Insight 1
top_10_medals = df_medals.sort_values(by='Total', ascending=False).head(10)

# What is .set_index()?
# - What it is: A method to set a DataFrame's index using one or more of its columns.
# - Why we use it: For plotting, it's often useful to have the labels (country names) as the index, which will then be used as the x-axis labels on our bar chart.
plot_data = top_10_medals.set_index('Team/NOC')

# What is .plot(kind='bar')?
# - What it is: Pandas DataFrames have a built-in .plot() method that acts as a wrapper around Matplotlib.
# - Why we use it: It provides a quick and convenient way to create plots directly from our data.
# - Parameters:
#   - kind='bar': Specifies that we want a bar chart.
#   - stacked=True: Stacks the bars for different columns (Gold, Silver, Bronze) on top of each other.
#   - figsize=(12, 7): Sets the size of the figure for better readability.
plot_data[['Gold', 'Silver', 'Bronze']].plot(kind='bar', stacked=True, figsize=(12, 7), 
                                             color=['gold', 'silver', '#cd7f32'])

# Add labels and title for clarity
plt.title('Medal Distribution for Top 10 Countries')
plt.ylabel('Number of Medals')
plt.xlabel('Country')
plt.xticks(rotation=45) # Rotate country names for better visibility
plt.legend(title='Medal Type')

# What is plt.show()?
# It displays the plot that we have constructed.
plt.show()

**✍️ Explanation of Insight 10:**

A stacked bar chart provides a much richer story than a simple table. We can instantly see not only the total medal count for each of the top 10 countries but also the composition of those medals. For example, we can visually confirm the United States' dominance in all three medal categories and see how Great Britain's high silver and bronze count contributes to its high total rank.

## Conclusion and Learning Summary

Congratulations on completing this assignment! You've gone from loading raw data to generating and visualizing complex insights. 

Along the way, we've learned and applied several core Pandas concepts:

- **Data Loading**: Using `pd.read_excel()` to bring data into a **DataFrame**.
- **Data Inspection**: Using `.head()` to quickly view our data.
- **Sorting and Ranking**: Using `.sort_values()` to find top performers.
- **Column Selection**: Selecting specific columns using `[['col1', 'col2']]` to focus our analysis.
- **Frequency Counting**: Using `.value_counts()` as a powerful shortcut for counting unique items.
- **Creating New Data**: Calculating new columns based on existing ones (e.g., `Female_Percentage`).
- **Grouping and Aggregating**: Using `.groupby()` to group data by category and perform calculations like `.size()`.
- **Merging DataFrames**: Using `pd.merge()` to combine different datasets based on a common key, the cornerstone of relational data analysis.
- **Handling Missing Data**: Using `.fillna()` to clean up our data after a merge.
- **Filtering**: Using boolean indexing (e.g., `df[df['column'] > 50]`) to isolate specific rows.
- **Plotting**: Using the integrated `.plot()` method and the `matplotlib` library to visualize our findings.

Pandas is a deep and powerful library, and this is just the beginning. By understanding these fundamental operations, you now have the foundational skills to tackle a wide variety of data analysis tasks.